nixos/modules/services/x11/extra-layouts.nix

```diff
@@ -0,0 +1,232 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  layouts = config.services.xserver.extraLayouts;
+
+  layoutOpts = {
+    options = {
+      description = mkOption {
+        type = types.str;
+        description = "A short description of the layout.";
+      };
+
+      languages = mkOption {
+        type = types.listOf types.str;
+        description =
+        ''
+          A list of languages provided by the layout.
+          (Use ISO 639-2 codes, for example: "eng" for english)
+        '';
+      };
+
+      compatFile = mkOption {
+        type = types.nullOr types.path;
+        default = null;
+        description = ''
+          The path to the xkb compat file.
+          This file sets the compatibility state, used to preserve
+          compatibility with xkb-unaware programs.
+          It must contain a <literal>xkb_compat { ... }</literal> block.
+        '';
+      };
+
+      geometryFile = mkOption {
+        type = types.nullOr types.path;
+        default = null;
+        description = ''
+          The path to the xkb geometry file.
+          This (completely optional) file describes the physical layout of
+          keyboard, which maybe be used by programs to depict it.
+          It must contain a <literal>xkb_geometry { ... }</literal> block.
+        '';
+      };
+
+      keycodesFile = mkOption {
+        type = types.nullOr types.path;
+        default = null;
+        description = ''
+          The path to the xkb keycodes file.
+          This file specifies the range and the interpretation of the raw
+          keycodes sent by the keyboard.
+          It must contain a <literal>xkb_keycodes { ... }</literal> block.
+        '';
+      };
+
+      symbolsFile = mkOption {
+        type = types.nullOr types.path;
+        default = null;
+        description = ''
+          The path to the xkb symbols file.
+          This is the most important file: it defines which symbol or action
+          maps to each key and must contain a
+          <literal>xkb_symbols { ... }</literal> block.
+        '';
+      };
+
+      typesFile = mkOption {
+        type = types.nullOr types.path;
+        default = null;
+        description = ''
+          The path to the xkb types file.
+          This file specifies the key types that can be associated with
+          the various keyboard keys.
+          It must contain a <literal>xkb_types { ... }</literal> block.
+        '';
+      };
+
+    };
+  };
+
+  patchIn = pkgs: name: layout: with layout;
+  ''
+      # install layout files
+      ${optionalString (compatFile   != null) "cp '${compatFile}'   'compat/${name}'"}
```

Comment:
And if you have the motivation, after another thought it would make a whole lot more sense for this patching to happen in the xkeyboardconfig derivation directly via an `extraSymbolFiles` argument you can `.override`.

Reply:
It sounds like a good idea but the expression for xkeyboardconfig is generated, I don't know if it's possible to add new arguments.

Reply:
How about adding a separate derivation `xkeyboardconfig_customizable` which directly takes these `extra*` arguments in https://github.com/NixOS/nixpkgs/blob/master/pkgs/servers/x11/xorg/overrides.nix

Something like
```nix
{
  xkeyboardconfig_customizable = { extraLayouts ? {} }:
    self.xkeyboardconfig.overrideAttrs ...;
}
```

Reply:
@infinisil do you another way to add options to an existing package? This does work but `xkeyboardconfig_customizable` will be a function instead of a derivation.

Reply:
I think this is fine, it's also how other attributes like this work

Reply:
Ok, then I think this is quite ready.